lahja/endpoint.py

```diff
@@ -61,36 +64,52 @@ def from_name(cls, name: str, base_path: Optional[pathlib.Path]=None) -> 'Connec
             raise TypeError("Provided `base_path` must be a directory")
 
 
-class EndpointConnector:
-    """
-    Expose the receiving queue of an :class:`~lahja.endpoint.Endpoint` so that any other
-    :class:`~lahja.endpoint.Endpoint` that wants to push events into this
-    :class:`~lahja.endpoint.Endpoint` can do so via the
-    :class:`~lahja.endpoint.ProxyEndpointConnector`.
-    """
+class Broadcast(NamedTuple):
+    event: Union[BaseEvent, bytes]
+    config: Optional[BroadcastConfig]
 
-    def __init__(self, endpoint: 'Endpoint') -> None:
-        self._endpoint = endpoint
 
-    def put_nowait(self, item_and_config: Tuple[BaseEvent, Optional[BroadcastConfig]]) -> None:
-        loop = self._endpoint._loop
-        if not self._endpoint._running:
-            self._endpoint.logger.warning(
-                "Attempted to push into %s while it isn't running.",
-                self._endpoint.name
-            )
-        # We need to wrap this in `call_soon_threadsafe` since otherwise, the event loop
-        # won't pick it up until some other task moves the loop forward
-        loop.call_soon_threadsafe(self._endpoint._receiving_queue.put_nowait, item_and_config)
+class RemoteEndpoint:
+    def __init__(self, reader: StreamReader, writer: StreamWriter) -> None:
+        self.writer = writer
+        self.reader = reader
+        self._drain_lock = asyncio.Lock()
 
+    @staticmethod
+    async def connect_to(path: str) -> 'RemoteEndpoint':
+        reader, writer = await asyncio.open_unix_connection(path)
+        return RemoteEndpoint(reader, writer)
 
-class ProxyEndpointConnector(BaseProxy):  # type: ignore # TypeShed is missing BaseProxy
-    """
-    Proxy that connects to an :class:`~lahja.endpoint.EndpointConnector`
-    """
+    async def send_message(self, message: Broadcast) -> None:
+        await _write_message(self.writer, message)
+        async with self._drain_lock:
+            # Use a lock to serialize drain() calls. Circumvents this bug:
+            # https://bugs.python.org/issue29930
+            await self.writer.drain()
 
-    def put_nowait(self, item_and_config: Tuple[BaseEvent, Optional[BroadcastConfig]]) -> None:
-        self._callmethod('put_nowait', (item_and_config,))
+    async def read_message(self) -> Broadcast:
+        return await _read_message(self.reader)
+
+    @property
+    def more_to_read(self) -> bool:
```

Comment:
Alternate name ideas.

- `has_more_to_read` 
- `is_reader_done`
- `is_reader_at_eof`  (most literal/explicit)
- `can_read`  (meh on this one)

Reply:
Looking at the call sites, I find `has_more_to_read` the most pleasant to use.

Reply:
I ended up going with `reader_not_at_eof`. I agree that `has_more_to_read` sounds nicest but I don't think I understand `at_eof` well enough to try to abstract it away behind a name, this is the name which makes the fewest promises.

Happy to discuss if you think something else is better!

Reply:
Sounds good! :+1: